### PR TITLE
[FLINK-7908][FLINK-7824][QS] Restructure QS packages and put QS jars in opt/.

### PR DESCRIPTION
**Pull request by kl0u**
## What is the purpose of the change

Make the Queryable State more usable by:
1) Reducing the dependencies the `client` jar (used by the user) transitively brings.
2) Putting the core and client jars in the `opt/` folder. From there, to activate QS the user has to put the core jar in the lib/ folder before starting the cluster, and use the client jar as a dependency to his/her program.

## Brief change log

Creating the `core` and `client` modules in the queryable state dir and moving classes around.
In addition, now the `client` module becomes a dependency of the `flink-runtime`.

Places the jars corresponding the aforementioned modules to the `opt` dir.

## Verifying this change

This change is a simple restructuring of the queryable state module, no classes are introduced or deleted.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes, adds the `flink-queryable-state-client` dependency to `flink-runtime`.

## Documentation

  - Does this pull request introduce a new feature? NO, but it needs documentation, which is PENDING.

R @aljoscha

### CODE REVIEW

File: flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/client/VoidNamespace.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+package org.apache.flink.queryablestate.client;
+
+import java.io.ObjectStreamException;
+
+/**
+ * Singleton placeholder class for state without a namespace.
+ *
+ * <p><b>THIS WAS COPIED FROM RUNTIME SO THAT WE AVOID THE DEPENDENCY.</b>
+ */
+public final class VoidNamespace {
```
Comment (aljoscha) [inline flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/client/VoidNamespace.java]: Can these be package private?

File: flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/FutureUtils.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.queryablestate;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Javadoc.
```
Comment (aljoscha) [inline flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/FutureUtils.java]: Missing Javadoc.

**Comment by aljoscha** [comment]
Btw, I'm assuming you also tried this with a cluster where you put the "core" jar in lib and then could query the state of a running job?

**Comment by aljoscha** [comment]
LGTM with the name changes!

Excellent work. 👍

**Comment by kl0u** [comment]
Thanks a lot for the review @aljoscha ! Waiting for travis and then merging.

**Comment by tillrohrmann** [comment]
Sorry for joining the party late, but I don't think that it is a good idea to duplicate code by copying code from one module into the other, as happened for the `FutureUtils` for example. The proper solution would be to move the `FutureUtils` into `flink-core` such that they are accessible from all places. Doing differently will lead us to maintenance hell quite quickly.

**Pull request merged**